File: 007.exs
# Juan Vazquez 
# https://github.com/javazquez
# http://javazquez.com
# Elixir Solution to Project Euler problem 7
# http://projecteuler.net/problem=7

defmodule Euler do
  def nextPrime(canidate,[h|t]) do
    if Enum.all?([h|t], &(rem(canidate,&1) !== 0 )) do
      [canidate|[h|t]] #prime so return it
    else
      nextPrime(canidate + 2, [h|t])
    end
  end
  def getNthPrime(x//10000) do #generator
    Stream.iterate([3], &(nextPrime(hd(&1)+2,&1))) |>
    Stream.take(x) |> 
    Enum.reverse |> 
    hd |> #[nth prime|[list of n primes]]
    hd #nth prime
  end
end

  Euler.getNthPrime |> 
  IO.puts
  #104743

File: 007.clj
;; Juan Vazquez
;; https://github.com/javazquez
;; http://javazquez.com
;; Clojure Solution to Project Euler problem 7
;; http://projecteuler.net/problem=7

(defn solve [ nth ]
  (loop [ canidate 3
          plist '(3 2 )
          acc 2
    ]
    (if (>= acc nth )
      (first plist)
      (if (prime? canidate plist)
        (recur (+ 2 canidate) (conj plist canidate) (inc acc))
        (recur (+ 2 canidate) plist   acc)))))

(defn prime? [ canidate plist]
  (->> plist
    (map #(rem canidate %) )
    (some zero?)
    not))

(time (solve 10001))
;"Elapsed time: 52900.907781 msecs"
;104743
